dash/development/base_component.py

```diff
@@ -387,6 +387,120 @@ def __repr__(self):
+    # TODO - Tab out the repr for the repr of these components to make it
+    # look more like a hierarchical tree
+    # TODO - Include "description" "defaultValue" in the repr and docstring
+    #
+    # TODO - Handle "required"
+    #
+    # TODO - How to handle user-given `null` values? I want to include
+    # an expanded docstring like Dropdown(value=None, id=None)
+    # but by templating in those None values, I have no way of knowing
+    # whether a property is None because the user explicitly wanted
+    # it to be `null` or whether that was just the default value.
+    # The solution might be to deal with default values better although
+    # not all component authors will supply those.
+    c = '''{docstring}
+
+    html{typename} <- function(..., {default_argtext}) {{
+
+    component <- list(
+      props = list(
+         {default_paramtext}
+      ),
+      type = '{typename}',
+      namespace = '{namespace}',
+      propNames = c({list_of_valid_keys}),
+      package = '{package_name}'
+    )
+
+    component$props <- filter_null(component$props)
+    component <- append_wildcard_props(component, wildcards = c({default_wildcards}), ...)
+
+    structure(component, class = c('dash_component', 'list'))
+    }}
+'''
+
+    filtered_props = reorder_props(filter_props(props))
+    # pylint: disable=unused-variable
+    list_of_valid_wildcard_attr_prefixes = repr(parse_wildcards(props))
+    # pylint: disable=unused-variable
+    list_of_valid_keys = repr(list(map(str, filtered_props.keys())))
+    list_of_valid_keys = list_of_valid_keys[1:-1]
+    package_name = make_package_name(namespace)
+    # pylint: disable=unused-variable
+    docstring = create_docstring_r(
+        component_name=typename,
+        props=filtered_props,
+        events=parse_events(props),
+        description=description).replace('\r\n', '\n')
+
+    # pylint: disable=unused-variable
+    # want to remove data-*, aria-*
+    events = '[' + ', '.join(parse_events(props)) + ']'
+    prop_keys = list(props.keys())
+    default_paramtext = ''
+    default_wildcards = ''
+    wildcard_list = ''
+    for key in props:
+        if '*' in key:
+            wildcard_list.join(key)
+    if 'children' in props:
+        prop_keys.remove('children')
+        default_argtext = "children=NULL, "
+        # pylint: disable=unused-variable
+        argtext = 'children=children, **args'
+    else:
+        default_argtext = ""
+        argtext = '**args'
+    # in R, we set parameters with no defaults to NULL, here we'll do that if no default value exists
+    default_wildcards += ", ".join(
+        [('\'{:s}\''.format(p))
+         for p in prop_keys
+         if '*' in p]
+    )
+    default_argtext += ", ".join(
+        [('{:s}={}'.format(p, props[p]['defaultValue']['value'])
+          if 'defaultValue' in props[p] else
+          '{:s}=NULL'.format(p))
+         for p in prop_keys
+         if not p.endswith("-*") and
+         p not in keyword.kwlist and
```

Comment:
:zap: This keyword list filter is for python keywords, won't be the same in R.

Reply:
Actually now we manually maintain a file of python keywords (https://github.com/plotly/dash/blob/master/dash/development/_all_keywords.py). Perhaps we can re-name that to python_keywords and make a new one called R_keywords.

That is, if this problem even exists in R. Can you define a function with argument names that are the same as a reserved keyword in R?

Reply:
> That is, if this problem even exists in R. Can you define a function with argument names that are the same as a reserved keyword in R?

If the function argument matches a reserved word, the R parser will throw an error:

`Error: unexpected 'if' in "myfun <- function(if"`

I've gone ahead and produced a GitHub gist including all the reserved words in R, similar to the keyword list already available for Python. I can submit it as a pull request if it would be helpful.

https://gist.github.com/rpkyle/622155e24f603fc54a664b06f8036f3d

Reply:
You can include those keywords in https://github.com/plotly/dash/blob/master/dash/development/_all_keywords.py as a new variable `r_keywords` and use it in this PR.

Reply:
Great, will do.

Reply:
I'm in the process of updating the code I've written, so please hold off on reviewing the updates to this PR for a day or so. Nicolas suggested remerging my new branch with R, so that we can keep the current PR alive.